rockcraft/extensions/flask.py

```diff
@@ -0,0 +1,169 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2023 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""An experimental extension for the Flask framework."""
+
+import copy
+import os
+import posixpath
+from typing import Any, Dict, Optional, Tuple
+
+from overrides import override
+
+from ._utils import _apply_extension_property
+from .extension import Extension
+
+
+class Flask(Extension):
+    """An extension for constructing Python applications based on the Flask framework."""
+
+    @staticmethod
+    @override
+    def get_supported_bases() -> Tuple[str, ...]:
+        """Return supported bases."""
+        return "bare", "ubuntu:20.04", "ubuntu:22.04"
+
+    @staticmethod
+    @override
+    def is_experimental(base: Optional[str]) -> bool:
+        """Check if the extension is in an experimental state."""
+        return True
+
+    @override
+    def get_root_snippet(self) -> Dict[str, Any]:
+        """Fill in some default root components for Flask.
+
+        Default values:
+          - run_user: _daemon_
+          - build-base: ubuntu:22.04 (only if user specify bare without a build-base)
+          - platform: amd64
+        """
+        snippet: Dict[str, Any] = {}
+        if "run_user" not in self.yaml_data:
+            snippet["run_user"] = "_daemon_"
+        if (
+            "build-base" not in self.yaml_data
+            and self.yaml_data.get("base", "bare") == "bare"
+        ):
+            snippet["build-base"] = "ubuntu:22.04"
+        if "platforms" not in self.yaml_data:
+            snippet["platforms"] = {"amd64": {}}
+        current_parts = copy.deepcopy(self.yaml_data.get("parts", {}))
+        current_parts.update(self._gen_new_parts())
+        snippet["parts"] = current_parts
+        snippet["services"] = self._gen_services()
+        return snippet
+
+    def _gen_services(self):
+        """Return the services snipped to be applied to the rockcraft file."""
+        services = {
+            "flask": {
+                "override": "replace",
+                "startup": "enabled",
+                "command": "/bin/python3 -m gunicorn --bind 0.0.0.0:8000 app:app",
+                "user": "_daemon_",
+                "working-dir": "/srv/flask/app",
+            }
+        }
+        existing_services = self.yaml_data.get("services", {})
```

Comment:
I think for consistency with the part handling this should also be a `deepcopy` of the existing data